File: internal/shadow/adapters/workflows.go
package adapters

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/stefanpenner-cs/reusable-workflows/internal/shadow/core"
)

// PatchWorkflowsInDir mirror-transforms the consumer's workflow files under
// <rootDir>/.github/workflows. Only files that actually call workflowsRepo are touched (leaving
// unrelated workflows alone so they aren't force-triggered on the shadow PR). Returns names changed.
func PatchWorkflowsInDir(rootDir, workflowsRepo, workflowsRef string) ([]string, error) {
	dir := filepath.Join(rootDir, ".github", "workflows")
	entries, err := os.ReadDir(dir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var changed []string
	for _, e := range entries {
		name := e.Name()
		if !strings.HasSuffix(name, ".yaml") && !strings.HasSuffix(name, ".yml") {
			continue
		}
		file := filepath.Join(dir, name)
		before, err := os.ReadFile(file)
		if err != nil {
			return nil, err
		}
		refs, err := core.ReferencesWorkflowsRepo(string(before), workflowsRepo)
		if err != nil {
			return nil, err
		}
		if !refs {
			continue
		}
		after, err := core.TransformWorkflowFile(string(before), workflowsRepo, workflowsRef)
		if err != nil {
			return nil, err
		}
		if after != string(before) {
			if err := os.WriteFile(file, []byte(after), 0o644); err != nil {
				return nil, err
			}
			changed = append(changed, name)
		}
	}
	return changed, nil
}

File: shadow/src/adapters/workflows.ts
import { readdirSync, readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { transformWorkflowFile } from '../core/transformWorkflowFile.ts';
import { referencesWorkflowsRepo, type PatchOptions } from '../core/patchConsumerWorkflow.ts';

/** Mirror-transform the consumer's workflow files under `<rootDir>/.github/workflows`. Only files
 * that actually call the workflows are touched — leaving unrelated workflows (e.g. a deploy job)
 * alone so they aren't force-triggered on the shadow PR. Returns the names of files changed. */
export function patchWorkflowsInDir(rootDir: string, opts: PatchOptions): string[] {
  const dir = join(rootDir, '.github', 'workflows');
  let names: string[];
  try {
    names = readdirSync(dir);
  } catch {
    return [];
  }

  const changed: string[] = [];
  for (const name of names) {
    if (!/\.ya?ml$/.test(name)) continue;
    const file = join(dir, name);
    const before = readFileSync(file, 'utf8');
    if (!referencesWorkflowsRepo(before, opts.workflowsRepo)) continue;
    const after = transformWorkflowFile(before, opts);
    if (after !== before) {
      writeFileSync(file, after);
      changed.push(name);
    }
  }
  return changed;
}
